tmt/steps/execute/__init__.py

```diff
@@ -362,6 +363,16 @@ def handle_reboot(self) -> bool:
             except ValueError:
                 timeout = None
 
+            flush_artifacts = reboot_data.get('flush_artifacts', flush_artifacts)
+
+            if flush_artifacts:
+                # Pull artifacts created in the plan data directory
+                self.logger.debug("Flush artifacts requested, pull the test data directory.", level=2)
+                if flush_func:
+                    flush_func()
+                else:
+                    self.guest.pull(source=self.test_data_path)
```

Comment:
1. the execution plugin already does pull stuff from the guest if allowed by its state (https://github.com/teemtee/tmt/blob/main/tmt/steps/execute/internal.py#L433) - would it make sense to add a pull of test data right there as well, to be executed every time by default?
2. I don't understand the difference between the default `flush_func`, which pulls the test data directory, but then this function is not used when called from the execution plugin, and the plan data directory pull is explicitly requested. IIUIC, the default value is never used in the current code.
   The "pull test data" comment is then quite misleading, because 1. the `flush_func` can do anything, it can erase my harddrive, and 2. it already does pull something else than the test data directory. Maybe we could just pass the path to pull instead of a callback, but then again, why do docs and default speak about test data but the code pulls plan data every time?
3. the name - I for one would recommend finding a name based on "pull", because it's already used by existing code and naming schemes, tmt "pulls" data from guests, it does not "flush" anything, plus "flush" suggests the subject is removed or sent somewhere rather than downloaded to where tmt runs.

Reply:
Thank you for the feedback

1. https://github.com/teemtee/tmt/blob/40ec859781af385ee88d4ba99f8441c4ac580439/tmt/steps/execute/internal.py#L433 in case of the system being restarted is too late, as it may not even get to that point id the system is lost (as described in the PR) as the reboot is happening here: https://github.com/teemtee/tmt/blob/40ec859781af385ee88d4ba99f8441c4ac580439/tmt/steps/execute/internal.py#L533 so the pull needs to be done before or while `invocation.handle_restart` is called, after that, it's too late.
2. Sorry about that confusion, I wasn't sure which solution is better one so I've put it to 2 commits (please check them individually). If the pull of the whole plan (not individual test) was desired, I'd drop completely the functionality introduced in the first commit.
3. You're right, from the perception of the test it's kind of force push (as the test is passing the instruction - request to pull the data from itself). I'm ok changing all the semantics to "pull" except of the `tmt-reboot` command where I believe "pull artifacts" could be super confusing.

Reply:
> Thank you for the feedback
> 
> 1. https://github.com/teemtee/tmt/blob/40ec859781af385ee88d4ba99f8441c4ac580439/tmt/steps/execute/internal.py#L433
>     in case of the system being restarted is too late, as it may not even get to that point id the system is lost (as described in the PR) as the reboot is happening here: https://github.com/teemtee/tmt/blob/40ec859781af385ee88d4ba99f8441c4ac580439/tmt/steps/execute/internal.py#L533
>     so the pull needs to be done before or while `invocation.handle_restart` is called, after that, it's too late.

Actually, this seems to be much less important after I had time to think about it: that `guest.pull()`, https://github.com/teemtee/tmt/blob/40ec859781af385ee88d4ba99f8441c4ac580439/tmt/steps/execute/internal.py#L433, happens right after the test finishes, way before reboot/restart are handled, and downloads the whole `invocation.path`. The test data directory is a subdirectory of `invocation.path`. So, the data you are after should be downloaded at that very point. The data are not downloaded when 1. watchdog check marked the guest as possibly dead and requested a hard reboot, or 2. the tests quit with an exit code it listed as a reason to be restarted. Neither of these two conditions should apply when test runs `tmt-reboot`, therefore the data should be downloaded.

One more thing:

> ... is suboptimal as the remaining (plan and other tests) data remain on the host and are risked to be lost

What "remaining data" are we speaking about? Data for previous tests should have been already pulled from the guest - if the current test crashes horribly and burns the guest to ashes, it shouldn't matter as previous tests finished their tasks before the current one, and tmt should have pulled their data to the safety of the tmt runner, before even starting the current test. And the follow-up tests produced no data on the guest yet. If you lack the data from previous tests, then again, something is horribly wrong.

The more I think about it, the more I get the impression there is something I don't understand, maybe a bug or more in tmt core. Something does not fit the picture I have. Because what you described and implemented should be a thing already. So either I misunderstand your use case, or I don't understand the code, or there is a bug somewhere and needs to be fixed. I'd like to find out which one it is :) Do you have a reproducer I could dissect and play with? That would help a lot.

Reply:
Sorry for the delay, had some other priorities. I've managed to prepare an example that shows what I'm trying to address: https://github.com/pholica/tmt_testrepo/blob/main/plans/example.fmf
When executed with the current TMT code, the only thing I get is `log.txt` and nothing more as the moment when the artifacts are being fetched is when the system is not alive.

Reply:
Hmm, ok, so I was playing with it little bit more and it really seems that the test data are pulled from the system after each test is finished and when `tmt-reboot` is invoked (makes sense as it also contains the reboot request) so what's not pulled is actually just the plan data which we currently heavily use due to the way how the log files are organized (by TMT) and presented by TF.

Reply:
Funny thing is that `tmt report` just (for some reason) removes those pulled files and claims that nothing was executed...

Reply:
And running it in TF also lead into a situation when these pulled artifacts are simply ignored. So I guess some other pieces needs to be touched, correct? I'm still thinking about this being relevant as the plan data are not pulled, but we could be fine by using test data instead...

Reply:
> Sorry for the delay, had some other priorities. I've managed to prepare an example that shows what I'm trying to address: https://github.com/pholica/tmt_testrepo/blob/main/plans/example.fmf When executed with the current TMT code, the only thing I get is `log.txt` and nothing more as the moment when the artifacts are being fetched is when the system is not alive.

```yaml
/tests:
  /01_log_plan_data:
    test: echo foobar > ${TMT_PLAN_DATA}/plan_log.txt
```

Indeed, this is probably lacking, because plan data are pulled just once, at the (successful) end of everything. I don't see any reason why not pulling this data after each test, but maybe there is a good reason against it. Will check with wiser heads.

```yaml
  /02_log_test_data:
    test: echo foobar > ${TMT_TEST_DATA}/test_log.txt
```

This one should be pulled before `03_do_damage_and_reboot` even starts, and it should really be available on the host on which tmt runs.

```yaml
  /03_do_damage_and_reboot:
    test: |
      rm -f /boot/vmlinuz*
      ls /boot > ${TMT_PLAN_DATA}/ls_boot.txt
      ls /boot > ${TMT_TEST_DATA}/ls_boot.txt
      tmt-reboot -t 300
    duration: 10m
```

Neither of these two can be available, once the guest is dead, there is no chance for tmt to pull them from the guest.

> Funny thing is that `tmt report` just (for some reason) removes those pulled files and claims that nothing was executed...

Might be because the `execute` step did not finish successfully, maybe with an exception and therefore it's still marked as `todo`. But I'm not 100 % sure of this explanation, needs investigation to confirm this reasoning.

> And running it in TF also lead into a situation when these pulled artifacts are simply ignored.

Ignored by whom, TF when rendering results page? Or by tmt, not pulling them?

> So I guess some other pieces needs to be touched, correct? I'm still thinking about this being relevant as the plan data are not pulled, but we could be fine by using test data instead...

It looks like the pull of plan data after each test would improve your situation, i.e. performing https://github.com/teemtee/tmt/blob/main/tmt/steps/execute/internal.py#L613 here https://github.com/teemtee/tmt/blob/main/tmt/steps/execute/internal.py#L432 rather than after all tests are done. @psss @lukaszachy what do you think, is it just a historic thing, or is there a technical or other reason against pulling plan data after each test? Seems harmless to me...

Reply:
> Neither of these two can be available, once the guest is dead, there is no chance for tmt to pull them from the guest.
Well, I can see the `${TMT_TEST_DATA}/ls_boot.txt` being pulled as it's being pulled along with the reboot request.
> Ignored by whom, TF when rendering results page? Or by tmt, not pulling them?
Ignored by TF, unfortunately I have link only to Red Hat ranch result page where I could show it, but if you run the test, you'll see how it's presented in TF.

Reply:
> > Neither of these two can be available, once the guest is dead, there is no chance for tmt to pull them from the guest.
> 
> Well, I can see the `${TMT_TEST_DATA}/ls_boot.txt` being pulled as it's being pulled along with the reboot request.

Ah, mea culpa, it's not a dead guest, but `tmt-reboot`. Then yes, the file would be pulled in.

Reply:
Added to hacking meeting agenda, to get some inputs.

Reply:
Ok, I found the issue, here's the fix for not saving the results: https://github.com/teemtee/tmt/pull/2996

Reply:
Still, after the fix in #2996 I see random behavior when fetching the test data from the tests prior to the one with the problematic reboot. In ~30% of cases, the test data from the `02_log_test_data` is present :/

Reply:
And I just found minimal reproducer for a very scary thing due to which I'm not getting the data from `02_log_test_data` https://github.com/teemtee/tmt/issues/2997

Reply:
@pholica seems #2997 was fixed